File: src/item.py
class Item:
    def __init__(self, src: str, chr: str):
        """ src is a valid string, if it contains only chr symbols and 4 '\n' symbols """
        self._posi, self._posj = 0, 0
        self.chr = chr
        self._valid = (src.count('\n') == 4 and src.count('#') == 4 and src.count('.') == 12)
        self._data = self._to_matrix(src.replace('#', chr))
        self.points = self._set_points()
        self._validate()

    def __str__(self):
        if self._data is None:
            return ""
        matrix = [['.' for _ in range(4)] for _ in range(4)]
        for i, j in self.points:
            matrix[i][j] = self.chr
        return "\n".join("".join(matrix[i]) for i in range(4))

    def _to_matrix(self, src: str):
        if self._valid:
            matrix = [line for line in src.split('\n')][:-1]
            for line in matrix:
                if len(line) != 4:
                    self._valid = False
                    return None
            return matrix
        return None

    def _get_offset(self):
        dj = min([line.find(self.chr) for line in self._data], key=lambda a: 3 if a < 0 else a)
        di = 0
        while not self._data[di].count(self.chr):
            di += 1
        return di, dj

    def _set_points(self):
        if self._data is None:
            return None
        di, dj = self._get_offset()
        points = set()
        for i in range(4):
            for j in range(4):
                if self._data[i][j] in self.chr:
                    points.add((i - di, j - dj))
        return points

    def _validate(self):
        """ check figure correctness """
        if self._data is None:
            return
        touchings = 0
        for i, j in self.points:
            if j - 1 >= 0 and (i, j - 1) in self.points:
                touchings += 1
            if j + 1 < 4 and (i, j + 1) in self.points:
                touchings += 1
            if i - 1 >= 0 and (i - 1, j) in self.points:
                touchings += 1
            if i + 1 < 4 and (i + 1, j) in self.points:
                touchings += 1
        if touchings < 6:
            self._valid = False
        else:
            self._valid = True

    def is_valid(self):
        return self._valid

    def set_position(self, i, j):
        self._posi, self._posj = i, j
    
    def get_position(self):
        return self._posi, self._posj

if __name__ == '__main__':
    A = ".###\n...#\n....\n....\n"
    i = Item(A, 'A')
    print(i)
    print(i.points)
    print("is correct?", i.is_valid())

File: src/C/item.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   item.c                                             :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: sjacelyn <sjacelyn@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/03/09 09:38:02 by sjacelyn          #+#    #+#             */
/*   Updated: 2019/03/09 19:10:52 by sjacelyn         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "item.h"

static char		**to_matrix(const char *s)
{
	char	**matrix;
	int		i;

	if (!(matrix = ft_strsplit(s, '\n')))
		return (NULL);
	i = 0;
	while (matrix[i])
	{
		if (ft_strlen(matrix[i]) != 4)
		{
			i = -1;
			while (matrix[++i])
				free(matrix[i]);
			free(matrix);
			return (NULL);
		}
		++i;
	}
	return (matrix);
}

static t_point	get_offset(char **matrix)
{
	t_point	offset;
	int		i;
	char	*ptr;

	offset.j = 4;
	i = 0;
	while (matrix[i])
	{
		ptr = ft_strchr(matrix[i], '#');
		if (ptr && ptr - matrix[i] < offset.j)
			offset.j = ptr - matrix[i];
		++i;
	}
	offset.i = 0;
	i = 0;
	while (matrix[i] && !ft_strchr(matrix[i], '#'))
	{
		offset.i = i;
		++i;
	}
	return (offset);
}

static void		set_points(t_item *item, char **matrix)
{
	t_point	offset;
	int		i;
	int		k;

	if (!matrix)
	{
		item->is_valid = 0;
		return ;
	}
	offset = get_offset(matrix);
	k = 0;
	i = 0;
	while (i < 16)
	{
		if (matrix[i / 4][i % 4] == '#')
		{
			item->points[k].i = i / 4 - offset.i;
			item->points[k].j = i % 4 - offset.j;
			++k;
		}
		++i;
	}
	while (--k >= 0)
		free(matrix[k]);
	free(matrix);
}

static void		validate_item(t_item *item)
{
	t_point	cur;
	t_point	other;
	int		c;
	int		i;
	int		j;

	if (!item->is_valid)
		return ;
	c = 0;
	i = -1;
	while (++i < 4)
	{
		cur = item->points[i];
		j = -1;
		while (++j < 4)
		{
			other = item->points[j];
			c += (cur.j - 1 >= 0 && other.i == cur.i && other.j == cur.j - 1);
			c += (cur.i - 1 >= 0 && other.i == cur.i - 1 && other.j == cur.j);
			c += (cur.j + 1 < 4 && other.i == cur.i && other.j == cur.j + 1);
			c += (cur.i + 1 < 4 && other.i == cur.i + 1 && other.j == cur.j);
		}
	}
	if (c < 6)
		item->is_valid = 0;
}

t_item			*create_item(char chr, const char *src)
{
	t_item	*item;

	if (!(item = malloc(sizeof(t_item))))
		return (NULL);
	item->chr = chr;
	item->posi = 0;
	item->posj = 0;
	item->is_valid = (ft_strlen(src) == 20 && ft_strcount(src, '\n') == 4 &&
					ft_strcount(src, '#') == 4 && ft_strcount(src, '.') == 12);
	set_points(item, to_matrix(src));
	validate_item(item);
	return (item);
}
